src/platform/Linux/MdnsImpl.cpp

```diff
@@ -0,0 +1,665 @@
+        VerifyOrExit(avahi_entry_group_add_service_strlst(mGroup, interface, AVAHI_PROTO_UNSPEC, static_cast<AvahiPublishFlags>(0),
+                                                          service.mName, type.c_str(), nullptr, nullptr, service.mPort, text) == 0,
+                     error = CHIP_ERROR_INTERNAL);
+    }
+    else
+    {
+        text = MakeAvahiStringListFromTextEntries(service.mTextEntryies, service.mTextEntrySize);
+
+        VerifyOrExit(avahi_entry_group_update_service_txt_strlst(mGroup, interface, AVAHI_PROTO_UNSPEC,
+                                                                 static_cast<AvahiPublishFlags>(0), service.mName, type.c_str(),
+                                                                 nullptr, text) == 0,
+                     error = CHIP_ERROR_INTERNAL);
+    }
+    VerifyOrExit(avahi_entry_group_commit(mGroup) == 0, error = CHIP_ERROR_INTERNAL);
+
+exit:
+    if (text != nullptr)
+    {
+        avahi_string_list_free(text);
+    }
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(DeviceLayer, "Avahi publish service failed: %d", static_cast<int>(error));
+    }
+
+    return error;
+}
+
+CHIP_ERROR MdnsAvahi::StopPublish()
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    VerifyOrExit(avahi_entry_group_reset(mGroup) == 0, error = CHIP_ERROR_INTERNAL);
+    VerifyOrExit(avahi_entry_group_commit(mGroup) == 0, error = CHIP_ERROR_INTERNAL);
+exit:
+    return error;
+}
+
+CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, chip::Inet::InterfaceId interface,
+                             MdnsBrowseCallback callback, void * context)
+{
+    AvahiServiceBrowser * browser;
+    std::shared_ptr<BrowseContext> browseContext = std::make_shared<BrowseContext>();
+
+    browseContext->mInstance = this;
+    browseContext->mContext  = context;
+    browseContext->mCallback = callback;
+    if (interface == INET_NULL_INTERFACEID)
+    {
+        interface = AVAHI_IF_UNSPEC;
+    }
+    browser = avahi_service_browser_new(mClient, interface, AVAHI_PROTO_UNSPEC, GetFullType(type, protocol).c_str(), nullptr,
+                                        static_cast<AvahiLookupFlags>(0), HandleBrowse, browseContext.get());
+    if (browser != nullptr)
+    {
+        mBrowseContexts[browseContext.get()] = browseContext;
+    }
+
+    return browser == nullptr ? CHIP_ERROR_INTERNAL : CHIP_NO_ERROR;
+}
+
+MdnsServiceProtocol TruncateProtocolInType(char * type)
+{
+    char * deliminator           = strrchr(type, '.');
+    MdnsServiceProtocol protocol = MdnsServiceProtocol::kMdnsProtocolUnknown;
+
+    if (deliminator != NULL)
+    {
+        if (strcmp("._tcp", deliminator) == 0)
+        {
+            protocol     = MdnsServiceProtocol::kMdnsProtocolTcp;
+            *deliminator = 0;
+        }
+        else if (strcmp("._udp", deliminator) == 0)
+        {
+            protocol     = MdnsServiceProtocol::kMdnsProtocolUdp;
+            *deliminator = 0;
+        }
+    }
```

Comment:
What if it's neither?

Reply:
It returns MdnsServiceProtocol::kMdnsProtocolUnknown as specified in the initialization.